docs/puzzles/day9.md

```diff
@@ -6,6 +6,154 @@ import Solver from "../../../../website/src/components/Solver.js"
+cell in the grid which contains a value smaller than the four adjacent values
+(up, down, left, right).
+
+I model the two dimensional grid containing height values using a case class
+called `Heightmap`. I provide it with a utility method which returns the list of
+coordinates and height values of the cells adjacent to a given cell. The list
+can have size 2, 3 or 4 depending whether the cell is on the edge of the grid or
+not:
+
+```scala
+type Height = Int
+case class Position(x: Int, y: Int)
+
+case class Heightmap(width: Int, height: Int, data: Vector[Vector[Height]]):
+
+  def apply(pos: Position): Height = data(pos.y)(pos.x)
+
+  def neighborsOf(pos: Position): List[(Position, Height)] =
+    val Position(x, y) = pos
+    List(
+      Option.when(x > 0)(Position(x - 1, y)),
+      Option.when(x < width - 1)(Position(x + 1, y)),
+      Option.when(y > 0)(Position(x, y - 1)),
+      Option.when(y < height - 1)(Position(x, y + 1))
+    ).flatMap(List.from)
+     .map(pos => pos -> apply(pos))
+```
+
+Using this method I implement the `lowPointsPositions` method which iterates over
+all the cells in the grid and for each of them, it checks if the value in it is
+smaller than the values of adjacent cells:
+
+```scala
+  def lowPointsPositions: LazyList[Position] =
+    LazyList.range(0, height).flatMap { y =>
+      LazyList.range(0, width).map { x => 
+        val pos = Position(x, y)
+        (
+          apply(pos),
+          pos,
+          this.neighborsOf(pos).map(_._2)
+        )
+      }
+    }
+    .collect {
+      case (value, poss, neighbors) if neighbors.forall(value < _) => 
+        poss
+    }
+```
+
+The method `forall` on collections returns true only if the predicate it is
+provided with holds true for all elements of the collection. It is similar to
+the `exists` methods which returns true if at least one element in the
+collection satisfies the provided predicate.
+Finally, `collect` allows us to simplify chains of `map` and `filter`.
+
+You can have a look at the documentation of these methods [in the
+documentation](https://www.scala-lang.org/api/current/scala/collection/immutable/Iterable.html)
+
+We now have all the tools to solve part 1:
+
+```scala
+def part1(input: String): Int =
+  val heightMap = Heightmap.fromString(input)
+
+  heightMap.lowPointsPositions.map(heightMap(_) + 1).sum
+end part1
+```
+
+<Solver puzzle="day9-part1"/>
+
+## Solution of Part 2
+
+In part 2 we have to create, for each low point, a basin: a group of grid cells
+that the smoke traverses while flowing downward towards the low point.
+
+This is how we can solve part 2 after implementing the basin creation for a single low point:
+
+```scala
+  val lowPoints = heightMap.lowPointsPositions
+  val basins = lowPoints.map(basin(_, heightMap))
+
+  basins
+    .to(LazyList)
+    .map(_.size)
+    .sorted(Ordering[Int].reverse)
+    .take(3)
+    .product
+```
+
+To build a basin from a low point I use very useful data structure provided in
```

Comment:
```suggestion
To build a basin from a low point I use a very useful data structure provided in
```
but perhaps "very useful" is not very useful in that sentence ;) Consider removing it.